### ISSUE TITLE
Add RISC-V vectors support

### ISSUE THREAD

**Issue by stalkerg** [libjpeg-turbo/libjpeg-turbo#620]
The RISC-V vectors are finally stable, and even dev boards are realized. I suppose it's a perfect time to start working on RISC-V support. 

Spec is here https://github.com/riscv/riscv-v-spec/releases/tag/v1.0
Here you can check examples: https://github.com/riscv/riscv-v-spec/blob/master/vector-examples.adoc

Regards,

**Comment by dcommander** [comment]
LOL. Yeah, as if I have the budget for that. This won’t happen without outside funding.

**Comment by stalkerg** [comment]
Yep, RISC-V Foundation probably can fund it, or somebody from SiFive
@kasanovic ?

**Comment by dcommander** [comment]
BTW, I feel I should provide more information about the general strategy behind adopting SIMD extensions in libjpeg-turbo.

In general, support for a new SIMD instruction set in libjpeg-turbo is driven by demand from application and operating system developers.  A complete set of libjpeg-turbo SIMD extensions can take hundreds of hours, so it's an expensive proposition.  Even if someone else does the initial implementation, it can still sometimes take anywhere from 30 to more than 100 hours for me to clean up the implementation, benchmark it, and integrate it, depending on the quality of the code.  The willingness of organizations to fund that labor is an imperfect measure of demand for the new instruction set.  Irrespective of funding, however, I am not interested in supporting CPU architectures that have little likelihood of adoption in major downstream software projects (such as popular web browsers) or end-user hardware products.

Additional concerns include the design of the SIMD instruction set and the CPU and compiler architecture.  My experience with RISC architectures is that, in general, the speedup to be gained from using SIMD instructions is less than that of CISC architectures, because the libjpeg-turbo scalar (non-SIMD) code is already faster with RISC CPUs and compilers.  (With PowerPC, for instance, I was able to get a 2x speedup with some difficulty, but certainly not the 6x speedup that was achievable with x86.)  My experience is also that not all SIMD instruction sets are created equal, and some of them are poor fits for the libjpeg-turbo algorithms.  (With 256-bit AVX2, for instance, some of the DCT algorithms simply couldn't be accelerated relative to 128-bit SSE2, because of the dual-lane nature of AVX2 and the fact that the DCT algorithms operate on 128 bits of data at a time.)  Other issues include whether compilers have a well-optimized implementation of compiler intrinsics for the instruction set.  (This bit us in the butt with Arm Neon.)  The RISC-V vector ISA is interesting in that it supports arbitrary vector lengths, but that raises additional concerns regarding whether a specific RISC-V SIMD implementation in libjpeg-turbo will provide consistent performance gains across all RISC-V CPUs or whether it will be necessary to handle differing capabilities in specific RISC-V CPU implementations.

Even if initial funding is provided to implement or integrate a new set of SIMD extensions, no set of SIMD extensions is zero-maintenance.  Once the initial implementation or integration is complete, then it is necessary to use the libjpeg-turbo General Fund to fix bugs in the implementation that are inevitably discovered down the road.  Thus, I must ensure that the platform is strategically important enough to adopt as a first-class citizen within the libjpeg-turbo ecosystem.  As much as RISC-V appeals to me, given my background in CPU architecture, I am frankly not convinced that it is strategically important enough yet.  It's a moot point at the moment, though, because I won't have any availability to work on such a project until well into 2023 at the earliest, even if funding is secured.

**Comment by stalkerg** [comment]
Thank you for the detailed explanation. I appreciate it.
Is it maybe good text even for a blog post?

**Comment by rednoah91** [comment]
Hi @dcommander , I am Hong-Rong from SiFive. We are interested in upstream RISC-V support in libjpeg-turbo. May I have your email to talk about the funding and co-working model?

**Comment by stalkerg** [comment]
@rednoah91 @dcommander I hope your collaboration will be a success. 
btw C implementation seems like working but because gcc anyway doesn't fully vectors support it's not really good.

**Comment by dcommander** [comment]
> Hi @dcommander , I am Hong-Rong from SiFive. We are interested in upstream RISC-V support in libjpeg-turbo. May I have your email to talk about the funding and co-working model?

@rednoah91 https://libjpeg-turbo.org/About/Contact

**Comment by BHbean** [comment]
actually I used to implement the SIMD version of libjpeg-turbo based on RVV1.0, which is my graduation project. It has already been open sourced: https://github.com/isrc-cas/libjpeg-turbo/tree/riscv-dev.
This implementation can already make sure the correctness of libjpeg-turbo, and the number of instructions on QEMU has significantly reduced. But I dont have the real machine at that time so I did not test it on a real system and left the implementation aside.
Recently I'm reviewing my previous work and find a possible chance to test it on a real machine. I think my previous work can help a lot, though it still has a lot needs to be improved.  How can I help you guys? @dcommander @rednoah91

**Comment by dcommander** [comment]
If your implementation is high-quality, then it may reduce the amount of labor I have to do in order to implement the feature, but that labor will still not be zero. All of my labor on this, even if it involves validating and integrating someone else's implementation, must be funded. Funding and access to test hardware are the two primary gating issues from my point of view.

**Comment by rednoah91** [comment]
Sorry for the late update. I am proposing RISE(RISC-V Software Ecosystem) to support the funding. It may take some time. I'll keep you posted.

**Comment by BHbean** [comment]
Thanks for you guys' update. I will spend my spare time to keep maintaining and updating the `riscv-dev` branch on the above repo. Hopefully my codes can do a little help for you.

**Comment by BHbean** [comment]
hey guys, i have some updates for RVV support. I add support for the newest rvv intrinsic version(v0.12 and v0.11), and test it on a real machine (Kendryte K230). The related document can be seen [here](https://github.com/isrc-cas/libjpeg-turbo/blob/riscv-dev/simd/rvv/README.md).

So how is the RISE project going now? @rednoah91 

And i notice that most of the tests in the test suite are executed very quickly, which makes me uncertain about performance improvement brought by my implementation. Is there other test codes to prove the performance enhencement? @dcommander

**Comment by dcommander** [comment]
@BHbean The benchmark procedure for new SIMD extensions is rather tedious.  Roughly speaking, what I do is run `tjbench {image} 95 -rgb -qq -nowrite -accuratedct -warmup 10` on the five test images described in [this report](https://libjpeg-turbo.org/About/Performance).  However, I want to get the results of those tests to within 1% accuracy, so I disable any kind of CPU speed throttling and all unnecessary processes on the (Linux) test system.  (Basically all that's running on the system are kernel daemons and an SSH server.)  I first establish a baseline performance reading with the latest stable release of libjpeg-turbo, which has none of the proposed SIMD extensions.  Then I start adding new SIMD extensions and benchmarking each one in isolation, in the order in which the algorithms appear on [this page](https://libjpeg-turbo.org/About/SIMDCoverage).  I start with the compression algorithms, so RGB-to-YCbCr and RGB-to-Grayscale Color Conversion are first.  Since those are compression algorithms, I expect the decompression speedup to be within +/- 1% relative to the baseline.  If it isn't, then I may need to increase the benchmark time (using the `-benchtime` argument to `tjbench`) on both the baseline and RGB-to-YCbCr/RGB-to-Grayscale benchmark runs until I can reliably get the decompression speedup to within the desired margin of error.  That gives me confidence that any compression speedup I observe is legitimate and not noise.  For each compression algorithm in sequence, I establish a new baseline in that manner, with the expectation being that each algorithm will speed up overall compression by significantly more than 1% but that decompression will remain within +/- 1% of the previous baseline.  If the speedup of a particular algorithm isn't what I expect, then I have to analyze the SIMD code carefully and figure out why it isn't performing as well as it should.  (It isn't unusual for a particular algorithm to require several development and benchmark passes.)  Once I have completed that procedure for all of the compression algorithms, then I have a new baseline that takes into account the total compression speedup, and I can use that new baseline to start testing decompression algorithms in sequence.  For the decompression algorithms, I expect the compression speedup to be within +/- 1% relative to the baseline, and I expect that each algorithm will speed up overall decompression by significantly more than 1%.  If I want to improve a particular SIMD extension after a full benchmark run is completed, then I benchmark that improvement relative to the final results of the benchmark run.  (It isn't necessary to redo the whole benchmark run.)  If SIMD extensions are added for the "fast" FDCT and IDCT algorithms, then the whole procedure must be repeated, replacing `-accuratedct` with `-fastdct` in the `tjbench` command line.  That means obtaining a new overall baseline with the reference (unmodified) release of libjpeg-turbo.  However, it is only necessary to benchmark the FDCT and IDCT algorithms in isolation, since the other algorithms were already benchmarked with the accurate FDCT/IDCT.  Similarly, if a SIMD extension is added for the progressive Huffman encoding algorithm, then the whole procedure must be repeated, adding `-progressive` to the `tjbench` command line, but only the progressive Huffman encoding algorithm must be benchmarked in isolation.  If SIMD extensions are added for the H2V1 and H2V2 merged upsampling algorithms, then the whole procedure must be repeated, adding `-fastupsample` to the `tjbench` command line, but only the merged upsampling algorithms must be benchmarked in isolation.

As you can imagine, this is extremely time consuming and requires some intuition regarding the behavior of the various algorithms and how much speedup should be expected from each.  This is why I don't generally accept contributions of SIMD extensions unless there is funding for me to peer review the extensions, which involves performing the aforementioned procedure.  This procedure is necessary in order to ensure that all proposed new SIMD extensions are justified.  (It wouldn't make sense to introduce a SIMD extension for a particular algorithm if that SIMD extension doesn't significantly speed up that algorithm.)  However, if the developer of the SIMD extensions can perform the benchmarking procedure during the development process and provide me with that data, then I will only have to peer review the final results, which takes significantly less time.  (This is assuming that there is a high degree of confidence that the data are reflective of the expected speedups across a wide array of CPUs within the architecture family, as well as a wide array of compilers and operating systems.  I mention that because, in the case of the Arm Neon SIMD extensions, it was necessary to separately benchmark in-order and out-of-order cores.  It was also necessary to separately benchmark GCC and Clang, because GCC < 12 didn't have an optimal Neon intrinsics implementation.)

This [OpenOffice Calc template](https://github.com/libjpeg-turbo/libjpeg-turbo/files/13604103/libjpegturbo-template.ods) may be useful.

NOTE: It might be possible to develop a lower-level benchmark that runs each algorithm in isolation repeatedly for a specified period of time, thus eliminating at least some of the aforementioned differential performance analysis tedium.  However, such a benchmark would likely be unrealistically cache-resident and thus wouldn't accurately reflect the performance of a particular SIMD extension in a real-world compression/decompression scenario.  (Also, you couldn't test the Huffman algorithms using such a low-level benchmark, because the performance of those algorithms strongly depends on the input.)  Another possibility might be to instrument the libjpeg-turbo code so that it independently tracks the average execution time of each algorithm for a particular libjpeg instance and prints a performance table when the instance is destroyed.  I'm not sure if that would really accelerate the development process, because it would still be desirable to benchmark each SIMD extension in isolation as it is developed.  It would, however, accelerate the peer review process, i.e. the process of benchmarking a set of SIMD extensions that has already been developed.

**Comment by BHbean** [comment]
It worked! Thanks a lot!
And two more questions (really sorry to bother a lot):
- `-accuratedct` option seems unavailable in the `tjbench-static`, can i take it as the default option?
- When i test the compression process, i cannot limit the speedup of decompression within +/- 1% no matter how i increase the value of `-benchtime`, should i disable the simd optimization for decompression to test the compression?

**Comment by dcommander** [comment]
> It worked! Thanks a lot! And two more questions (really sorry to bother a lot):
> 
> * `-accuratedct` option seems unavailable in the `tjbench-static`, can i take it as the default option?
> * When i test the compression process, i cannot limit the speedup of decompression within +/- 1% no matter how i increase the value of `-benchtime`, should i disable the simd optimization for decompression to test the compression?

`-accuratedct` is the default in libjpeg-turbo 3.0.  The spreadsheet template was developed for prior versions, in which it was necessary to specify `-accuratedct`.  I thought that the `-accuratedct` option was still available but hidden, but apparently I removed it altogether.  Sorry about that.

The procedure I described above is a procedure for validating the performance of new SIMD extensions as you develop them, so it assumes that the decompression algorithms won't be SIMD-accelerated when you're developing and benchmarking the compression SIMD extensions.  If you can get the decompression performance to within +/- 1% by disabling the decompression SIMD extensions, then yes, that's what you should do when benchmarking the compression SIMD extensions.  It would also be valid to disable the compression SIMD extensions while benchmarking the decompression SIMD extensions, even though my procedure and the spreadsheet template don't do that.

It does concern me, however, that the SIMD performance is so variable.

**Comment by stalkerg** [comment]
@BHbean just in case, do you know any new boards with final vector instructions? I want to buy one, because in qemu we mostly can validate code itself but not a performance. Also, it will be good to understand our real targets.

**Comment by BHbean** [comment]
@stalkerg Kendryte K230 was used to be available on [Taobao](https://m.tb.cn/h.5Ozd0cC?tk=yweAWUqnWgj), a Chinese e-shop platform. However, it has been sold out now. So maybe there are no available boards with RVV 1.0 now. I heard that SiFive might be developing such products, so maybe you could wait for a while :)

**Comment by dcommander** [comment]
I'm surprised that there are still no public boards available. If that is the case, then it is too early to consider this technology for inclusion in libjpeg-turbo. SIMD extensions for libjpeg-turbo potentially take hundreds of hours to implement the right way. My concern is that I will finish implementing them (or validating and integrating someone else's implementation) only to find that the implementation performs optimally on the pre-release hardware and not on the final hardware. I speak from experience in that regard, since the differences in various Arm implementations were a source of pain for our project in the past. Even if someone pays me for the initial implementation or integration work, I will ultimately have to maintain the feature using General Fund money, just like I maintain every other set of SIMD extensions in libjpeg-turbo. Thus, I have a strong desire to not develop against a moving target. I may have said something similar in the comments above, and if so, I apologize for repeating myself. I was just under the impression that the technology was farther along than it apparently is.

**Comment by rayeh5** [comment]
We (Andes Technology) implemented the IDCT functions on source files listed as below:
simd/rvv/jidctint-rvv.c, simd/rvv/jidctred-rvv.c and simd/rvv/jidctfst-rvv.c.
The RISC-V Vector Intrinsic version used is v0.11 and RVV veriosn 1.0.
The verification is done on Andes’ FPGA and chip platform.
We think it can be helpful for the RVV porting and will contribute if you need.

**Comment by stalkerg** [comment]
@rayeh5, for now, it will be better if you can make PR to the @BHbean branch.

> I'm surprised that there are still no public boards available.

@dcommander, it's available now and will be available more in 2024; the situation is not so bad, but yes, it's early stage.

**Comment by BHbean** [comment]
> @rayeh5, for now, it will be better if you can make PR to the @BHbean branch.

@stalkerg Thanks for your mentioning! My branch welcome every commit that brings pratical  performance improvement. BTW, Kendryte K230 has been [resaled](https://detail.tmall.com/item.htm?abbucket=2&id=743909802431) now, maybe you could have a look :)

> We (Andes Technology) implemented the IDCT functions on source files listed as below:
> simd/rvv/jidctint-rvv.c, simd/rvv/jidctred-rvv.c and simd/rvv/jidctfst-rvv.c.
> The RISC-V Vector Intrinsic version used is v0.11 and RVV veriosn 1.0.

@rayeh5 I am considering make the `libjpeg-turbo-rvv` an independent repo for now, since contributing to the upstream on a RISC-V arch does not seem to be accepted now. And I also implement `simd/rvv/jidctint-rvv.c` and `simd/rvv/jidctfst-rvv.c` in my [repo](https://github.com/isrc-cas/libjpeg-turbo/tree/riscv-dev), maybe you can compare the two versions with each other on your FPGA and chips platform. It would be an honor for me to see my implemention running on other platforms rather than K230, since I am really pleased and curious about the performance over different platforms. You can open an issue [here](https://github.com/isrc-cas/libjpeg-turbo/tree/riscv-dev) and we can disscus more details there.

**Comment by dcommander** [comment]
I am in discussions with several key people regarding funding for my labor to complete the implementation, which will involve reviewing any existing code and also introducing a new SIMD profiling framework into libjpeg-turbo to facilitate that review. I do not accept piecemeal SIMD extensions, because it is much more efficient to review and test a complete implementation.

At this point, I would encourage you to be patient rather than fork libjpeg-turbo for the purpose of introducing an incomplete feature. I have never received a pull request for SIMD extensions that didn't require extensive modifications on my part, so I am understandably skeptical of other people's SIMD code until it has been thoroughly reviewed. I am happy to discuss the implementation in general terms, but funding needs to be secured before I can do much else.

**Comment by stalkerg** [comment]
@dcommander as I understand @BHbean doesn't do a project fork, he just maintain a repo fork to accumulate changes for rvv support. It's can be already helpful for some people and he can get a good feedback. To provide you a finished PR people should develop it somewhere and test it.

**Comment by chLFF** [comment]
> @BHbean just in case, do you know any new boards with final vector instructions? I want to buy one, because in qemu we mostly can validate code itself but not a performance. Also, it will be good to understand our real targets.

Banana Pi with SpacemiT K1 may satisfy.
https://docs.banana-pi.org/en/BPI-F3/SpacemiT_K1_datasheet

**Comment by stalkerg** [comment]
This year, will be at least 4 devices with RVV 1.0, mostly come from SiFive chip design. We already have some work in forks, next year should be mature enough to talk about it here.
(I am checking the risc-v state each months)

**Comment by dcommander** [comment]
From my end, I was prepared to begin work on this in January, and I thought that the wheels were in motion to get a contract in place for me to do so. However, the people I was corresponding with in email (from the Linux Foundation, Google, and SiFive) haven't responded to me since January. (I just re-pinged them last week and still haven't heard back.) I need to be looped in on the design of this feature to make sure that it will be acceptable to our project once completed. There are some basic technical questions that haven't yet been answered (such as whether it is possible to implement the DCT algorithms using variable-length vectors, etc.) Historically, SIMD extensions developed externally to this project have always required costly and time-consuming modifications before they could be integrated.

**Comment by stalkerg** [comment]
> such as whether it is possible to implement the DCT algorithms using variable-length vectors, etc

Seems like possible:
https://code.videolan.org/videolan/dav1d/-/merge_requests/1463/diffs

**Comment by dcommander** [comment]
Do not assume that it is possible for libjpeg-turbo just because it is possible for other software. Due to the constraints of the libjpeg API and architecture, our DCT/IDCT algorithms must work on only one 8x8 DCT block at a time. Because the algorithms have inter-row and inter-column dependencies, that means that they can only operate on 128 bits of data at a time (8 x 16-bit coefficients.) The only reason I was able to take advantage of 256-bit AVX2 instructions is that the accurate DCT/IDCT algorithms use 32-bit intermediates for the 16-bit inputs and outputs. None of the other SIMD implementations use vectors larger than 128-bit at the moment. I posit that variable vector lengths could be used for color conversion and maybe down/upsampling but not for any of the frequency domain stuff.

**Comment by dcommander** [comment]
To reiterate the situation:  I am willing and able to begin work on this project, but since RISC-V doesn't benefit me or my other OSS projects, I am unwilling to work on it unless all of my labor is funded.  As of January 2024, I was in discussions with contacts at SiFive, Google, and The Linux Foundation regarding funding and access to a development board, but they stopped replying to my emails.  I tried pinging them again in April but still heard nothing, so I gave up.  At this point, I am not optimistic that I will ever hear anything.  In general, with features like this that have little or no existing demand within the libjpeg-turbo community, funded development is a litmus test for whether the feature can amass a significant user base in the future.  If no one is willing to invest the (relatively small, in corporate terms) amount of funding required for me to implement or integrate a feature, then that signals to me that the technology may not be ready for inclusion in a JPEG reference implementation.  If someone wants to prove me wrong on that, then all they have to do is secure enough funding for me to implement or integrate this feature.  Otherwise, I don't earn a salary for my work on libjpeg-turbo, and the libjpeg-turbo General Fund (monthly donations provided by a couple of corporations, as well as a dozen or so individual GitHub sponsors) pays for only a day or two of labor a month, which is rarely enough to even maintain the project and fix critical bugs.

**Comment by camel-cdr** [comment]
RISE has started a ["RISE RISC-V Developers Appreciation Pilot Program"](https://github.com/Rise-dev-appreciation/Rise-dev-appreciation), although this project would require more funding and more stable funding, than submit after the project is done.

@luhenry maybe you can help work something out?

**Comment by dcommander** [comment]
@camel-cdr The effort required would minimally fall into the Large category, but if I have to do a compete implementation from scratch, it would exceed that level of funding.  (A complete implementation could take 100-200 hours, based on my past experience with Arm and PowerPC.)

**Comment by stalkerg** [comment]
Small update - we finally have a cheap dev board with RVV1 support https://www.amazon.com/dp/B0DZ6VWRSZ

**Comment by BHbean** [comment]
@filipwasil Thanks a lot for your efforts in collecting all the updates made for RVV arch, I am so happy that someone still cares about this issue. I fully support the idea that we as RISE members should maintain a separate repo for all the progress in RVV, which truly demonstrates the spirit of collaboration in open-source community.

As for my own codebase, I received PRs from Andes before and merged one into my branch. I think @rayeh5 (from Andes too) has made his efforts in upstreaming changes to my branch, and you can check the [PRs](https://github.com/ISRC-CAS/libjpeg-turbo/pulls?q=) here for details.

**Comment by dcommander** [comment]
> [@filipwasil](https://github.com/filipwasil) Thanks a lot for your efforts in collecting all the updates made for RVV arch, I am so happy that someone still cares about this issue. I fully support the idea that we as RISE members should maintain a separate repo for all the progress in RVV, which truly demonstrates the spirit of collaboration in open-source community.

I still care about it, but I'm also still waiting after two years for those who proposed funding for labor and equipment to follow through on their proposals. Collaboration is a lot easier when you're a full-time employee with a guaranteed salary. It is less easy when you're an independent developer living month to month. There are tradeoffs to both approaches. Developing libjpeg-turbo independently and funding it through patronage and funded development frees me from any one company's agenda. (If I were a corporate employee, it is unlikely that I would be able to work on RISC-V at all.) Developing it as more of a "cathedral" than a "bazaar" allows me to maintain high levels of quality in the code base without the need for project management processes or dedicated teams that are typically needed for quality control in a collaborative environment. However, that approach also precludes me from accepting most code contributions as-is, and it precludes me from working on new features unless my labor is funded. You want to integrate with libjpeg-turbo because libjpeg-turbo is stable, performant, ubiquitous, cross-platform, and an industry standard. That is a result of how the project is managed, but my inability to integrate or develop RISC-V extensions for free is also a result of how the project is managed.

As a former computer architect, I like how RISC-V and RVV are designed and would actually love to work on RVV extensions for libjpeg-turbo, but I can't move on the project until/unless funding is secured.  For now, I agree that collecting and stabilizing the existing libjpeg-turbo/RVV code in a downstream repository might be the best approach.  That may at least reduce the eventual upstream integration labor, although upstream integration will still have to happen eventually.

[This comment](https://github.com/libjpeg-turbo/libjpeg-turbo/pull/640#issuecomment-1380341121) under the AVX512 feature proposal may provide some insight into how I approach SIMD extension development, which may help steer downstream development in a direction that minimizes the eventual upstream integration labor.  tl;dr: Make sure that all contributions are documented in the copyright headers of the modified/new files, make sure that each SIMD algorithm is implemented in a separate commit, and make sure that each SIMD algorithm in isolation significantly increases the overall performance (which I typically measure using [this procedure](https://libjpeg-turbo.org/About/Performance).)

My more specific concern with RVV is whether the DCT/IDCT algorithms can support multiple vector lengths.  Those algorithms operate on 8-deep rows and columns of 16-bit coefficients with interdependencies between each row/column that prevent multiple rows/columns from being computed in parallel.  Thus, it isn't generally possible to take advantage of vectors larger than 128 bits in the "fast" DCT/IDCT algorithms, and the "slow" DCT/IDCT algorithms can only take advantage of 256-bit vectors because they use 32-bit values for intermediate computations.

The good news is that libjpeg-turbo is being awarded a grant through FLOSS/fund.  Once that money is secured, then it will open up the possibility for me to work on this without direct funding.  However, there will still be a limit to how much time I can spend on it, so the more you can do to reduce my eventual development/integration labor, the better.

**Comment by camel-cdr** [comment]
@dcommander I don't think the libjpeg-turbo design currently supports this, and I don't think this is what initial support should look like, but you should be able to vectorize the DCT/IDCT scalably, by just doing VLEN/128 independant DCTs/IDCTs in parallel.

I'm currently experimenting with this using the stb_image.h codebase, which is smaller and easier to hack on, and I'll share my results here once/if I got it working.

**Comment by dcommander** [comment]
> [@dcommander](https://github.com/dcommander) I don't think the libjpeg-turbo design currently supports this, and I don't think this is what initial support should look like, but you should be able to vectorize the DCT/IDCT scalably, by just doing VLEN/128 independant DCTs/IDCTs in parallel.
> 
> I'm currently experimenting with this using the stb_image.h codebase, which is smaller and easier to hack on, and I'll share my results here once/if I got it working.

I suspected as much regarding VLEN/128, but what do you mean regarding doing independent DCTs/IDCTs in parallel?

**Comment by dcommander** [comment]
> For the 8x8 IDCT you have 8 vector registers that each store a single row of the matrix. For VLEN>=128 you can instead store VLEN/128 (e.g. for VLEN=512: VLEN/128=512/128=4) rows of seperate matricies in the vector registers.

How would this work given the computational interdependencies between rows/columns and the limitations of the libjpeg API architecture?  Referring to the discussion under the aforementioned AVX-512 feature proposal, this has come up in the past, both in the context of a failed effort to add OpenCL SIMD extensions to libjpeg-turbo and also a failed effort to accelerate the "fast" DCT/IDCT algorithms using AVX2 instructions.  Because the libjpeg API allows calling applications to finely control the amount of data that is processed with each function call, I assert that taking advantage of vector lengths > 128 bits in the DCT/IDCT algorithms would require providing a baseline 8x8 version of the DCT/IDCT algorithm as well as multi-block versions (16x8, 32x8, etc.) that allow multiple DCT blocks to be processed in one throw. However, that would require additional methods and logic in both the coefficient controller and the DCT/IDCT managers, which I have been reluctant to attempt. (It's not that it's impossible. It's just that, in my experience, making deep changes to a 35-year-old code base and architecture creates numerous unforeseen issues down the road. That's why the SIMD extension architecture in libjpeg-turbo is designed to literally replace all of the low-level C methods on a 1:1 basis with exactly the same inputs and outputs.)  If you have a way to leverage 256-bit or 512-bit instructions on a single 8x8 block in isolation, then that would be hugely beneficial, since it would open the door to further acceleration with AVX2, AVX-512, SVE, etc.

**Comment by dcommander** [comment]
libjpeg-turbo has been awarded a large grant through FLOSS/fund, which I intend to use in part for SIMD enhancements.  Thus, if there is any current work on RVV extensions for libjpeg-turbo, please point me to it.

**Comment by filipwasil** [comment]
https://gitlab.com/riseproject/libjpeg-turbo/-/merge_requests/2

This is the merge request that we have prepared for our internal REPO waiting for a good moment to submit it to upstream (Which I believe is now). It is not merged yet. It can be rebased over the recent upstream with no conflicts (Tried it last week). It was tested on QEMU and banana pi f3 and reviewed internally. Nothing is perfect, but i suspect that this may be a pretty good start.

**Comment by stalkerg** [comment]
Maybe it's a little bit off topic here, but can we ask libjpeg folks to improve API to unlock a modern and wide SIMD and Vectors?

**Comment by camel-cdr** [comment]
I'm still looking into doing multiple IDCTs simultaneously, but I hadn't had much time to spend on it. I think I'm getting there and I'll report back with benchmarks once I have them.

One thing I did is measure different implementations for the IDCT transpose and found using masked vslide1up/vslide1down to be the fastest (until the the [vzip](https://riscv.atlassian.net/wiki/spaces/VXXX/pages/168886273/Fast-Track+Proposal+for+vector+zip+instructions) extension is ratified):

* [transpose 8x8 8-bit elements](https://camel-cdr.github.io/rvv-bench-results/bpi_f3/trans8x8e8.html)
* [transpose 8x8 16-bit elements](https://camel-cdr.github.io/rvv-bench-results/bpi_f3/trans8x8e16.html)
* [link to assembly implementation of vslide1up/vslide1down transpose](https://github.com/camel-cdr/rvv-bench/blob/65a8dec6f238a45f22b26d02638471bfe68461c5/bench/trans8x8.S.inc#L292)

You can also see that, at least for the transpose in isolation, there is a good amount of performance to be gained from working on multiple 8x8 matrices at once.

To save you some time, here are a few resources I recommend:

* [ISA manual](https://riscv.github.io/riscv-isa-manual/snapshot/unprivileged/), the [isolated "V" spec](https://github.com/riscv/riscv-isa-manual/blob/main/src/v-st-ext.adoc) for faster page loads
* [rvv intrinsics docs](https://github.com/riscv-non-isa/rvv-intrinsic-doc), but I recommend using [dzaimas intrinsics viewer](https://dzaima.github.io/intrinsics-viewer/)
* [c-api doc](https://github.com/riscv-non-isa/riscv-c-api-doc/blob/main/src/c-api.adoc)
* [RISE optimization guide](https://gitlab.com/riseproject/riscv-optimization-guide)
* [my rvv-bench project](https://camel-cdr.github.io/rvv-bench-results/index.html) tries to measure performance characteristics of existing RVV hardware

**Comment by dcommander** [comment]
> Maybe it's a little bit off topic here, but can we ask libjpeg folks to improve API to unlock a modern and wide SIMD and Vectors?

It's not off-topic at all.  Currently libjpeg-turbo can't fully utilize vector lengths > 128 bits, which limits the SIMD acceleration possibilities with AVX2 (this is why the "fast" integer forward/inverse DCTs have no AVX2 implementation), [AVX-512](https://github.com/libjpeg-turbo/libjpeg-turbo/pull/640), and RVV (since it supports multiple vector lengths, and implementations aren't required to support 128-bit vectors.)

The issue with using vector lengths > 128 bits isn't just in the API.  It's in the underlying libjpeg architecture, which dates back to 1991 (nearly 20 years before the existence of libjpeg-turbo.)  Most of the SIMD-accelerated algorithms work with image rows, and those algorithms can use any vector length.  The forward and inverse DCT algorithms, however, operate on one 8x8 coefficient block at a time.  Since there are serial dependencies between the rows/columns in the block, that means that only one 8x16-bit row or column can be processed at a time, which limits the ability to fully utilize vector lengths > 128.

The coefficient controller performs entropy encoding/decoding, forward/inverse DCT, sample conversion, and quantization on one 8x8 block at a time, but it has access to an entire "MCU row." (An MCU row is the width of the image and the height of one MCU.  Only the luma MCU will have multiple 8x8 blocks, and only if chroma subsampling is used.)  Thus, with a lot of effort, it would theoretically be possible for the coefficient controller to feed multiple blocks into the forward/inverse DCT algorithms.  Minimally, we would need special 16x8 (256-bit) and/or 32x8 (512-bit) forward/inverse DCT modules and a complete reorganization of the coefficient controller and DCT manager logic to accommodate those modules.  However, for optimal performance, it may be necessary to reorganize the multi-block data in memory so that adjoining rows or adjoining blocks are contiguous.  That would necessitate special 16x8 (256-bit) and/or 32x8 (512-bit) sample conversion and quantization modules as well as additional reorganization of the logic.  That's a lot of effort for probably not a lot of performance gain.

I would suggest initially implementing the RVV SIMD extensions so that the forward/inverse DCT modules support only 128-bit vectors. (Other modules can be implemented to support arbitrary vector lengths.)  The SIMD dispatcher should check the supported vector lengths at run time and avoid dispatching to the RVV forward/inverse DCT modules if 128-bit vectors aren't supported.  After that is done, then we'll have a solid baseline from which to evaluate whether the multi-block enhancements create enough of a performance boost to justify the disruption to our 35-year-old fragile codec architecture.

**Comment by dcommander** [comment]
Also, please rebase any proposed RVV implementation on the dev branch of libjpeg-turbo, which contains a complete overhaul of the SIMD dispatchers.

**Comment by filipwasil** [comment]
> Also, please rebase any proposed RVV implementation on the dev branch of libjpeg-turbo, which contains a complete overhaul of the SIMD dispatchers.

https://github.com/libjpeg-turbo/libjpeg-turbo/pull/837

**Comment by filipwasil** [comment]
I will rebase it

**Comment by dcommander** [comment]
Thanks.  I will acquire a board and start reviewing the PR as soon as the grant money lands.

**Comment by dcommander** [comment]
This feature has been cleaned up, documented, and significantly optimized from #837.  Refer to the [attached spreadsheet](https://github.com/user-attachments/files/25050210/libjpegturbo-3.2-riscv.ods) for detailed performance measurements, but the tl;dr is that I was able to improve upon the PR by ~10-26% (avg. 18%) for RGB-to-baseline compression and ~1-14% (avg. 8%) for baseline-to-RGB decompression.  Almost all of that came from optimizations to the color conversion, downsampling, and upsampling algorithms, which benefit the most from variable vector lengths.  Overall speedup (relative to C/scalar) is ~137-204% (avg. 173%) for compression and ~15-106% (avg. 65%) for decompression. Primary areas that are still suboptimal:

- The DCT and IDCT modules currently use strided loads/stores to transpose the 8x8 matrix in memory.  An efficient in-register transpose algorithm would probably speed things up considerably, but I was unable to implement one using RVV 1.0.  I tried moving the coefficients into a 16m1x8 tuple register group, using `*vrgather*()` to transpose the coefficients, and moving them back out into 16m1 registers, but that was slower than using strided loads/stores.  Using segmented loads/stores instead of strided loads/stores didn't improve performance significantly.  The accurate IDCT is only barely faster than C code at the moment.

    The main issue with implementing an efficient in-register matrix transpose seems to be the lack of a fast interleaving or permutation instruction.  Thus, the new Zvzip RVV extension may provide further opportunities for optimizing these algorithms.

    The current implementation of these algorithms (which I didn't change much relative to the PR) is almost a line-for-line port of my 128-bit AltiVec implementation.  There may be some opportunity for improving performance by reorganizing the algorithms around the strengths of RVV, but I feel like the most low-hanging fruit is an efficient in-register matrix transpose.

- The speedup of color deconversion is less than what I would expect, so there may be opportunities for improvement there, particularly if some of the new Zvzip instructions allow us to pack pixels in-register rather than relying on a segmented store.  I also noticed that the range-limiting instructions (`*vmax*()` and `*vmin*()`) are slow.  I was able to eliminate `*vmin*()` by using a saturating right shift, but `*vmax*()` is still necessary to clamp negative values to 0.  Other SIMD module implementations can get away with not doing that, so it might be fruitful to examine why.

- The fancy upsampling algorithms currently use non-continguous memory accesses to load the downsampled components and their nearest neighbors.  That could probably be improved by doing a single load operation with N + 1 samples, then using in-register operations to create the primary and nearest-neighbor vectors.  However, trying to figure out how to do that with a variable vector length broke my brain.  (The problem is that, because of tail handling, loading N+1 samples might necessitate an additional iteration of the outer loop if there are exactly VLMAX elements left to process.)
 
   Anyhow, the overall speedup of these algorithms is much better than what we get on x86 or Arm, so I don't consider this to be a high priority.

- My review comments under #837 were not answered, so I had to go it alone on figuring out how to clean up build-time and run-time RVV detection.  Please review my work on that.  One remaining issue is the use of GNU assembler (GAS) code to assist with run-time RVV 1.0 detection.  I would strongly prefer that function to be reimplemented using inline assembly in **simd/riscv64/jsimdcpu.c**.  libjpeg-turbo 3.2 is jettisoning all other GAS code.  (GAS was used only in the MIPS DSP r2 SIMD extensions, which are now EOL, and as fallbacks for older versions of GCC that lack optimal Arm Neon intrinsics support.)  In general, GAS code doesn't always assemble properly with Clang, so I would like to avoid it.

- The new low-level benchmarking feature (`WITH_PROFILE`) in the libjpeg-turbo 3.2 Evolving code base makes it a lot easier to examine the performance of SIMD modules in isolation.  I would encourage others to use that feature if they wish to explore some of the remaining performance issues, but bear in mind that (per Amdahl's Law) the speedup of any one algorithm needs to be really significant before it will affect overall performance by very much.  The aforementioned spreadsheet shows per-algorithm speedups on my fastest x86 and Arm test machines, for comparison.

    The primary focus should be on the DCT/IDCT algorithms, with secondary focus on color deconversion (including merged upsampling) and h2v2 fancy downsampling.  Those are the algorithms for which we currently see a lesser speedup than is achievable on x86 and Arm CPUs.

- Bonus points if someone can implement Huffman encoding using RVV, as Huffman encoding represents the biggest chunk of overall compresion performance.  It tends to be really hard to implement using SIMD instructions, however, because it is inherently a serial algorithm.

- This effort required nearly 100 hours of uncompensated labor on my part.  That was only possible because of a generous General Fund grant from FLOSS/fund, but this feature has put me in danger of exhausting that grant before the rest of the libjpeg-turbo 3.2 features can be implemented.  In fact, RVV support could've been implemented 2-3 years ago had the relevant players not ghosted me on funding discussions.  This is the only major SIMD implementation in libjpeg-turbo that did not receive any specific funding, and that fails to impress me.

    libjpeg-turbo is an ISO/ITU-T reference implementation, so I have to make sure that every new piece of code is as stable, well-implemented, and readable as the existing code.  That is an extremely demanding job.  I'm not trying to sound arrogant, but the mere fact that I have 16 years of experience with this project makes it unlikely that anyone else can do that job as well as I can.  In general, the biggest hurdle I face as libjpeg-turbo maintainer/principal developer is convincing people that the experience necessary to produce code suitable for a critical infrastructure project is worth something, and that the minimum necessary code to implement a feature will never be suitable.

    tl;dr: Don't expect me to devote hundreds more hours of free labor to this feature going forward.

- Feel free to discuss issues with the implementation here or under #837.  Please do not file additional PRs without discussing them with me first.

**Comment by camel-cdr** [comment]
Thanks, for the amazing work!

> The DCT and IDCT modules currently use strided loads/stores to transpose the 8x8 matrix in memory. An efficient in-register transpose algorithm would probably speed things up considerably, but I was unable to implement one using RVV 1.0.

There is a faster method with baseline RVV, which involves using even/odd masked vslide1down/up instructions. An mask-odd `vslide1up.vx` is `TRN1/vpaire.vv` and a mask-even `vslide1down.vx` is `TRN2/vpairo.vv`, this does require a few extra moves though.
Here is an [assembly implementation](https://github.com/camel-cdr/rvv-bench/blob/16b08a4bfa954e5b5a26e5206fbf911266be81a3/bench/trans8x8.S.inc#L279) and [intrinsic implementation](https://github.com/camel-cdr/rvv-playground/blob/2a154b3164a54cdbbb85c7f4a1e2bfd2803398db/stb_image/stb_image.h#L3092).
Note that gcc struggles a bit with register allocation, it spills if you look at the transpose code gen standalone, but integrated into the IDCT it doesn't.

>  I also noticed that the range-limiting instructions (*vmax*() and *vmin*()) are slow

I don't expect this to be the case in future hardware. The SpacemiT X60 had some wired quirks in general, e.g. using `vadd.vi vd, vs1, 0` has a 2x higher throughput than the regular move instructions `vmv.v.v` and `vmv1r.v`. But for now, we have to make due with current hardware.

> The fancy upsampling algorithms currently use non-continguous memory accesses to load the downsampled components and their nearest neighbors. That could probably be improved by doing a single load operation with N + 1 samples, then using in-register operations to create the primary and nearest-neighbor vectors. However, trying to figure out how to do that with a variable vector length broke my brain

I'd like to give it a try as well, once you've uploaded your changes.

> The primary focus should be on the DCT/IDCT algorithms, with secondary focus on color deconversion (including merged upsampling) and h2v2 fancy downsampling

So, regarding my comments from now almost a year ago.
After doing a bunch of other stuff, I recently finished my POC for doing multiple IDCTs simultaneously.
I've integrated it as a fast-path into stb_image.h, because that was easier to modify: https://github.com/camel-cdr/rvv-playground/blob/main/stb_image/stb_image.h (for now it's only in the non-progressive interleaved code path)

On the SpacemiT X60, the simultaneously IDCTs gave an additional 24% speedup in the total decompression time of a 4745x4756 9.4M image, which was compressed with a 90% compression ratio:

```
* 2.13 seconds: original
* 1.23 seconds: with stb_image.h RVV support (1 IDCT at a time with reduced LMUL)
* 0.93 seconds: with RVV support and fast-path (VLEN/128 IDCTs at a time)
```

The majority of the remaining time is spend in Huffman decoding, which is a lot trickier to vectorize. I have some ideas, but I'm not sure they will be profitable.

I just wanted to share this. While I wrote it for RVV, doing multiple IDCTs simultaneously could also be profitable for AVX and AVX-512. Presumably, this would require a rather restructuring of the library, so I'd see this as a separate issue/feature from general RVV support.
I'm also not sure how large of an improvement you can expect, as I suspect the majority of time in libjpeg-turbo is also spend in Huffman decoding.

**Comment by dcommander** [comment]
> There is a faster method with baseline RVV, which involves using even/odd masked vslide1down/up instructions. An mask-odd `vslide1up.vx` is `TRN1/vpaire.vv` and a mask-even `vslide1down.vx` is `TRN2/vpairo.vv`, this does require a few extra moves though. Here is an [assembly implementation](https://github.com/camel-cdr/rvv-bench/blob/16b08a4bfa954e5b5a26e5206fbf911266be81a3/bench/trans8x8.S.inc#L279) and [intrinsic implementation](https://github.com/camel-cdr/rvv-playground/blob/2a154b3164a54cdbbb85c7f4a1e2bfd2803398db/stb_image/stb_image.h#L3092). Note that gcc struggles a bit with register allocation, it spills if you look at the transpose code gen standalone, but integrated into the IDCT it doesn't.

Thanks for the tip.  I'll try to integrate it and re-benchmark in the next couple of days.

> > The fancy upsampling algorithms currently use non-continguous memory accesses to load the downsampled components and their nearest neighbors. That could probably be improved by doing a single load operation with N + 1 samples, then using in-register operations to create the primary and nearest-neighbor vectors. However, trying to figure out how to do that with a variable vector length broke my brain
> 
> I'd like to give it a try as well, once you've uploaded your changes.

Which changes?

> I just wanted to share this. While I wrote it for RVV, doing multiple IDCTs simultaneously could also be profitable for AVX and AVX-512. Presumably, this would require a rather restructuring of the library, so I'd see this as a separate issue/feature from general RVV support. I'm also not sure how large of an improvement you can expect, as I suspect the majority of time in libjpeg-turbo is also spend in Huffman decoding.

See my comment above regarding the challenges of restructuring the libjpeg architecture to accommodate this.  It's hypothetically doable, but it would easily get into hundreds of hours more labor just for the implementation alone.  Inevitably that will compound into more labor down the road, because once I change something that fundamental in a library that implements a 35-year-old extremely complex and exposed API, unexpected stuff will break.  The last time I made such a disruptive change was to implement partial image decompression 11 years ago, and issues have popped up with that even as recently as last year.

**Comment by camel-cdr** [comment]
> Which changes?

The RVV implementation and the changes you made to the PR. Unless I've missed it, they aren't on a public branch yet.

**Comment by dcommander** [comment]
They're in the dev branch.

**Comment by stalkerg** [comment]
@dcommander Thank you for your work! It's amazing what we finally in that point even without RISC-V companies support.
I am really happy what my ticket was closed. 
PS Seems like now we should push vendors to support Zvzip RVV ASAP.

**Comment by dcommander** [comment]
> There is a faster method with baseline RVV, which involves using even/odd masked vslide1down/up instructions. An mask-odd `vslide1up.vx` is `TRN1/vpaire.vv` and a mask-even `vslide1down.vx` is `TRN2/vpairo.vv`, this does require a few extra moves though. Here is an [assembly implementation](https://github.com/camel-cdr/rvv-bench/blob/16b08a4bfa954e5b5a26e5206fbf911266be81a3/bench/trans8x8.S.inc#L279) and [intrinsic implementation](https://github.com/camel-cdr/rvv-playground/blob/2a154b3164a54cdbbb85c7f4a1e2bfd2803398db/stb_image/stb_image.h#L3092). Note that gcc struggles a bit with register allocation, it spills if you look at the transpose code gen standalone, but integrated into the IDCT it doesn't.

I tried your code via the [attached patch](https://github.com/user-attachments/files/25071207/0001-RVV-In-register-transpose-experiment.patch).  It sped up the accurate integer inverse DCT by 5% and the fast integer inverse DCT by 10%, but both of the forward DCTs slowed down by 6%.  This is with GCC 14.2.0 on Ubuntu 24.04.

**Comment by dcommander** [comment]
> > > The fancy upsampling algorithms currently use non-continguous memory accesses to load the downsampled components and their nearest neighbors. That could probably be improved by doing a single load operation with N + 1 samples, then using in-register operations to create the primary and nearest-neighbor vectors. However, trying to figure out how to do that with a variable vector length broke my brain
> > 
> > 
> > I'd like to give it a try as well, once you've uploaded your changes.

I tried this as well, via the [attached patch](https://github.com/user-attachments/files/25075375/0001-RVV-Avoid-non-contiguous-memory-access-experiment.patch).  It actually made the fancy upsampling performance slightly worse.  I give up.  I have never encountered another platform in which memory accesses (especially non-continguous ones) are faster than in-register operations.

**Comment by dcommander** [comment]
@camel-cdr Would love your input on the comments above.

**Comment by camel-cdr** [comment]
> I tried your code via the [attached patch](https://github.com/user-attachments/files/25071207/0001-RVV-In-register-transpose-experiment.patch). It sped up the accurate integer inverse DCT by 5% and the fast integer inverse DCT by 10%, but both of the forward DCTs slowed down by 6%. This is with GCC 14.2.0 on Ubuntu 24.04.

@dcommander Have you tried the in-register-transpose patch with clang?

It looks like gcc spills in the codegen for `jsimd_fdct_islow_rvv` and `jsimd_idct_islow_rvv`: https://godbolt.org/z/9E6vqd7Y7
gcc also generates about 200 more instructions than clang, which seem to mostly be redundant `vmv1r.v` instructions, because gcc struggles with register allocation.
BTW both clang and gcc generate a csrr vlenb on spill, so you can easily check if something spilled by searching for "vlenb".

The iDCT range limit could be better using vmax/vmin, see modified [patch.txt](https://github.com/user-attachments/files/25276692/patch.txt).

I ran the current master branch, dev branch, and my patch above on the SpacemiT X100 (which I got early ssh access to):


```
build with clang-21 and gcc-15
./tjbench testimages/shira_bird8.bmp 95 -rgb -qq -nowrite -warmup 10 -benchtime 30 -dct int
./tjbench testimages/shira_bird8.bmp 95 -rgb -qq -nowrite -warmup 10 -benchtime 30 -dct fast

  master          dev           patch
clang gcc     clang gcc     clang gcc
46.78 49.58   66.54 71.25   69.88 71.76
5.116 5.116   5.116 5.116   5.116 5.116
77.44 74.16   78.59 79.87   83.50 81.29

31.05 33.40   48.05 51.71   50.70 52.03
4.477 4.477   4.477 4.477   4.477 4.477
44.82 42.00   55.45 56.22   60.28 57.53

25.12 27.12   39.50 42.78   41.88 43.01
4.102 4.102   4.102 4.102   4.102 4.102
41.24 39.14   46.58 47.40   51.02 48.56

19.39 21.41   30.07 32.96   31.85 33.03
3.600 3.600   3.600 3.600   3.600 3.600
34.68 31.14   36.69 37.50   40.47 38.49

-dct fast
55.65 53.13   71.81 75.81   74.39 73.53
5.086 5.086   5.086 5.086   5.086 5.086
81.39 79.13   85.23 85.48   91.44 89.20

37.26 35.80   52.25 55.29   54.48 53.85
4.468 4.468   4.468 4.468   4.468 4.468
46.43 44.57   61.54 60.31   66.43 63.67

30.62 29.35   43.68 46.31   45.50 44.76
4.096 4.096   4.096 4.096   4.096 4.096
43.01 42.33   52.38 51.77   57.11 54.66

24.42 23.53   33.42 36.05   35.08 34.78
3.588 3.588   3.588 3.588   3.588 3.588
36.20 34.02   41.78 41.51   46.31 44.42
```

> I tried this as well, via the [attached patch](https://github.com/user-attachments/files/25075375/0001-RVV-Avoid-non-contiguous-memory-access-experiment.patch). It actually made the fancy upsampling performance slightly worse. I give up. I have never encountered another platform in which memory accesses (especially non-continguous ones) are faster than in-register operations.

This is certainly weird. Maybe the loads can be more freely scheduled?

**Comment by dcommander** [comment]
Please use a bigger image than **shira_bird8.bmp** for benchmarking.  I tend to prefer **artificial.ppm** from [this data set](https://imagecompression.info/test_images/rgb8bit.zip), because it has a good mix of things that are hard and easy for the JPEG algorithms to compress.  However, really any suitably large image (generally > 5 Mpixels) is fine.

Referring to the [attached new results](https://github.com/user-attachments/files/25301067/libjpegturbo-3.2-riscv.ods), the data was a bit noisy, but it looks like the range limiting tweaks biased the low-level DCT/IDCT performance upward by a few percent.

You are correct that the in-register transpose shows a significant speedup across the board with Clang, significant enough to justify a few percent compression slow-down with GCC relative to my implementation-- particularly given the assumption that GCC will eventually improve its register allocation, as it did with the Neon intrinsics.  (The final compression results with GCC are still somewhat faster than the initial implementation in PR #837.)

The fancy upsampling tweak (avoiding non-contiguous memory accesses) didn't make a significant difference in performance, but I committed it anyway since it is better code.

All of the above are now in the dev branch.  The transpose algorithm is essentially the same as yours, but for readability, I unrolled it, added illustrative comments, and used the same intermediate vector names as in the AltiVec implementation.  I added your name to the copyright header of that file as well.

**Comment by camel-cdr** [comment]
Some small improvements:

```diff
diff --git a/simd/riscv64/jccolext-rvv.c b/simd/riscv64/jccolext-rvv.c
index 56b41b7..be6b1ff 100644
--- a/simd/riscv64/jccolext-rvv.c
+++ b/simd/riscv64/jccolext-rvv.c
@@ -109,3 +109,3 @@ jsimd_rgb_ycc_convert_rvv(JDIMENSION img_width, JSAMPARRAY input_buf,
       cb32 = __riscv_vwmulu_vx_u32m4(b16, F_0_500, vl);
-      cb32 = __riscv_vadd_vx_u32m4(cb32, SCALED_CENTERJSAMPLE, vl);
+      cb32 = __riscv_vadd_vx_u32m4(cb32, SCALED_CENTERJSAMPLE + ONE_HALF - 1, vl);
       tmp = __riscv_vwmulu_vx_u32m4(g16, F_0_331, vl);
@@ -115,3 +115,2 @@ jsimd_rgb_ycc_convert_rvv(JDIMENSION img_width, JSAMPARRAY input_buf,
       /* Narrow to 16-bit and round. */
-      cb32 = __riscv_vadd_vx_u32m4(cb32, ONE_HALF - 1, vl);
       cb16 = __riscv_vnsrl_wx_u16m2(cb32, SCALEBITS, vl);
@@ -122,3 +121,3 @@ jsimd_rgb_ycc_convert_rvv(JDIMENSION img_width, JSAMPARRAY input_buf,
       cr32 = __riscv_vwmulu_vx_u32m4(r16, F_0_500, vl);
-      cr32 = __riscv_vadd_vx_u32m4(cr32, SCALED_CENTERJSAMPLE, vl);
+      cr32 = __riscv_vadd_vx_u32m4(cr32, SCALED_CENTERJSAMPLE + ONE_HALF - 1, vl);
       tmp = __riscv_vwmulu_vx_u32m4(g16, F_0_418, vl);
@@ -128,3 +127,2 @@ jsimd_rgb_ycc_convert_rvv(JDIMENSION img_width, JSAMPARRAY input_buf,
       /* Narrow to 16-bit and round. */
-      cr32 = __riscv_vadd_vx_u32m4(cr32, ONE_HALF - 1, vl);
       cr16 = __riscv_vnsrl_wx_u16m2(cr32, SCALEBITS, vl);
```

```diff
diff --git a/simd/riscv64/jdcolext-rvv.c b/simd/riscv64/jdcolext-rvv.c
index 1587337..88df056 100644
--- a/simd/riscv64/jdcolext-rvv.c
+++ b/simd/riscv64/jdcolext-rvv.c
@@ -38,3 +38,3 @@ jsimd_ycc_rgb_convert_rvv(JDIMENSION out_width, JSAMPIMAGE input_buf,
   vuint8m2_t y, cb, cr;
-  vint16m4_t y16, cb16, cr16, r16, g16, b16;
+  vint16m4_t y16, cb16, cr16, r16, g16, b16, cb2;
   vint32m8_t g32;
@@ -105,3 +105,3 @@ jsimd_ycc_rgb_convert_rvv(JDIMENSION out_width, JSAMPIMAGE input_buf,
 
-      b16 = __riscv_vsll_vx_i16m4(cb16, 1, vl);                    /* 2 * Cb */
+      b16 = cb2 = __riscv_vsll_vx_i16m4(cb16, 1, vl);              /* 2 * Cb */
       b16 = __riscv_vmulh_vx_i16m4(b16, -F_0_228, vl);
@@ -111,4 +111,3 @@ jsimd_ycc_rgb_convert_rvv(JDIMENSION out_width, JSAMPIMAGE input_buf,
       b16 = __riscv_vsra_vx_i16m4(b16, 1, vl);           /* Cb * -FIX(0.228) */
-      b16 = __riscv_vadd_vv_i16m4(b16, cb16, vl);         /* Cb * FIX(0.772) */
-      b16 = __riscv_vadd_vv_i16m4(b16, cb16, vl);         /* Cb * FIX(1.772) */
+      b16 = __riscv_vadd_vv_i16m4(b16, cb2, vl);          /* Cb * FIX(1.772) */
       b16 = __riscv_vadd_vv_i16m4(b16, y16, vl);      /* Y + Cb * FIX(1.772) */
diff --git a/simd/riscv64/jdmrgext-rvv.c b/simd/riscv64/jdmrgext-rvv.c
index 8563360..645e464 100644
--- a/simd/riscv64/jdmrgext-rvv.c
+++ b/simd/riscv64/jdmrgext-rvv.c
@@ -39,3 +39,3 @@ jsimd_h2v1_merged_upsample_rvv(JDIMENSION output_width, JSAMPIMAGE input_buf,
   vuint8m1_t ye, yo, cb, cr;
-  vint16m2_t ye16, yo16, cb16, cr16, r_y16, g_y16, b_y16;
+  vint16m2_t ye16, yo16, cb16, cr16, r_y16, g_y16, b_y16, cb2;
   vint16m2_t re16, ro16, ge16, go16, be16, bo16;
@@ -132,3 +132,3 @@ jsimd_h2v1_merged_upsample_rvv(JDIMENSION output_width, JSAMPIMAGE input_buf,
     /* B-Y */
-    b_y16 = __riscv_vsll_vx_i16m2(cb16, 1, vl);                    /* 2 * Cb */
+    b_y16 = cb2 = __riscv_vsll_vx_i16m2(cb16, 1, vl);              /* 2 * Cb */
     b_y16 = __riscv_vmulh_vx_i16m2(b_y16, -F_0_228, vl);
@@ -138,4 +138,3 @@ jsimd_h2v1_merged_upsample_rvv(JDIMENSION output_width, JSAMPIMAGE input_buf,
     b_y16 = __riscv_vsra_vx_i16m2(b_y16, 1, vl);         /* Cb * -FIX(0.228) */
-    b_y16 = __riscv_vadd_vv_i16m2(b_y16, cb16, vl);       /* Cb * FIX(0.772) */
-    b_y16 = __riscv_vadd_vv_i16m2(b_y16, cb16, vl);       /* Cb * FIX(1.772) */
+    b_y16 = __riscv_vadd_vv_i16m2(b_y16, cb2, vl);        /* Cb * FIX(1.772) */
     /* Add Y */
```

**Comment by camel-cdr** [comment]
There should be a rather large speedup in the DCT code, if you have a dedicated VLEN>=256 implementation, that works on LMUL=1/2 for SEW=16 and LMUL=1 for SEW=32.
This is because the code can only do one 8x8 DCT at a time, so on VLEN>=256 only the lower 8 elements are used.
Some RVV implementations issue based on vl (a vl=1 operation performs the same for LMUL=1 and LMUL>=2). Unfortunately, most current ones only look at LMUL to determine execution, including the SpacemiT ones, so having one VLEN=128 and one VLEN>=256 implementation would improve performance by quite a lot.
I'll try to write this in the future.

**Comment by dcommander** [comment]
How would the VLEN>=256 implementation differ from the current implementation? Would it just operate on fractional registers?

**Comment by camel-cdr** [comment]
Yes, basically find and replace `i16m1` with `i16mf2` and `i32m2` with `i32m1`, adjusting all of the other stuff that changes.
There is a complication though, because LMUL=1/2 SEW=64 isn't allowed, for some reason, so the transpose could would need to reinterpret to/from LMUL=1 (vlmul_ext/vlmul_trunc) when SEW=64.

**Comment by dcommander** [comment]
> Yes, basically find and replace `i16m1` with `i16mf2` and `i32m2` with `i32m1`, adjusting all of the other stuff that changes. There is a complication though, because LMUL=1/2 SEW=64 isn't allowed, for some reason, so the transpose could would need to reinterpret to/from LMUL=1 (vlmul_ext/vlmul_trunc) when SEW=64.

If I have some time this week, I'll try that on one of the DCT/IDCT algorithms and see if it makes a difference on my machine.

**Comment by dcommander** [comment]
> There should be a rather large speedup in the DCT code, if you have a dedicated VLEN>=256 implementation, that works on LMUL=1/2 for SEW=16 and LMUL=1 for SEW=32. This is because the code can only do one 8x8 DCT at a time, so on VLEN>=256 only the lower 8 elements are used. Some RVV implementations issue based on vl (a vl=1 operation performs the same for LMUL=1 and LMUL>=2). Unfortunately, most current ones only look at LMUL to determine execution, including the SpacemiT ones, so having one VLEN=128 and one VLEN>=256 implementation would improve performance by quite a lot. I'll try to write this in the future.

That generally improved the low-level  accurate DCT/IDCT performance by 20-30%.  With Clang, it improved the "fast" DCT/IDCT performance by 14%.  With GCC, it improved the "fast" IDCT performance by 7% with negligible improvement for the "fast" forward DCT.  Changes will be pushed shortly.

**Comment by camel-cdr** [comment]
I had an idea:

```diff
diff --git a/src/jdhuff.c b/src/jdhuff.c
index cd8c0847..ace3f9ff 100644
--- a/src/jdhuff.c
+++ b/src/jdhuff.c
@@ -412,3 +412,31 @@ no_more_bytes:
 
+#if __riscv_v && !__riscv_misaligned_fast
+#define LOAD_U64(dst,ptr) __riscv_vmv_x(__riscv_vreinterpret_u64m1(__riscv_vle8_v_u8m1(ptr, 8)))
+#else
+#define LOAD_U64(dst,ptr) memcpy(&dst,ptr,8)
+#endif
+
+#if __riscv_zbb
+#define ANY_BYTE_ZERO(x) __riscv_orc_b_64(bits)
+#else
+#define ANY_BYTE_ZERO(x) ((((bits) - 0x0101010101010101) & ~(bits)) & 0x8080808080808080)
+#endif
+
 /* Pre-fetch 48 bytes, because the holding register is 64-bit */
+#if defined(__x86_64__) || defined(__aarch64__) || \
+    (__riscv_zbb && (__riscv_misaligned_fast || __riscv_v))
+#define FILL_BIT_BUFFER_FAST \
+  if (bits_left <= 16) { \
+    size_t bits; \
+    LOAD_U64(bits,buffer); \
+    bits = __builtin_bswap64(bits << 16); \
+    if (ANY_BYTE_ZERO(~bits)) { \
+      GET_BYTE GET_BYTE GET_BYTE GET_BYTE GET_BYTE GET_BYTE \
+    } else { \
+      get_buffer = (get_buffer << 48) | bits; \
+      bits_left += 48; \
+      buffer += 6; \
+    } \
+  }
+#else
 #define FILL_BIT_BUFFER_FAST \
@@ -417,2 +445,3 @@ no_more_bytes:
   }
+#endif
```

This adds a SWAR fast-path for huffman decode. Initially I tried to use vcompress to handle `0xFF`, but then I realized that encountering a `0xFF` in six bytes is rather rare, so I tried the fast-path.
I only tested this on x86 so far and it seems to give a slight 1-2% speedup, but it's hard to measure.

The code would currently not be used in RISC-V, but Ubuntu 26.04 enables RVA23 by default, so the prepackaged gcc would compile with `-march=rv64gcbv_zvbb...` automatically and this code path exposed.

This might also be something to adjust in the `simd/CMakeLists.txt`, instead of overwriting `march`. It would probably be better, to check if the default compiler builds a program with RVV intrisics, if that fails try it again with `-march=rv64gcv`.
This should get better codegen on Ubuntu 26.04 and other distros that raise the baseline, because the compiler could also use the bitmanip extensions.

**Comment by dcommander** [comment]
- Any changes that affect x86-64 or AArch64 would have to be much more carefully considered and tested on a variety of architectures.
- I am generally not interested in changes that affect overall performance by 1-2%.
    - It is very difficult to make TJBench measure that tight of a margin.  I generally have to shut down all non-essential processes on the (Linux) machine, lock the benchmark process to one core, run it for a fairly long period of time (long enough to make it painful to perform multiple benchmark runs), and benchmark a changed section of code along with an unchanged section (e.g. benchmark a change in the compressor while benchmarking the unchanged decompressor or vice versa.)  Then I have to repeat the benchmark until the unchanged section of code shows < 1% variation, which gives me a sliver (but only a sliver) of confidence that any > 1% variation in the changed section of code is legitimate.
    - A 1-2% speedup is not going to be noticed by end users, so it is generally not worth the risk of code disruption or reducing readability and maintainability.
    - My experience is that, when a change produces that small of a speedup with one platform/compiler, there's a good chance that it will produce a slowdown with another platform/compiler.
    - My experience is that you can often produce the same speedup by tweaking compiler settings without disrupting the code.
 - I've already spent more than 100 hours of labor on this, which was only possible because of a generous grant from FLOSS/fund, but that grant has been exhausted by this and other major libjpeg-turbo 3.2 features.  The current RVV SIMD modules are already 25-45% faster on average than the initial implementation that was submitted in PR #837.  From a technical point of view, it was an interesting and worthwhile effort.  From a project management and business point of view, I need to move on from it unless the corporate beneficiaries of the effort are willing to provide funding for more improvements.
 - By default, the build system only passes `-march=rv64gcv` to the compiler when building the RVV SIMD modules.  That is to allow for building libjpeg-turbo in such a way that it can perform run-time RVV detection and thus run on machines with or without RVV 1.0 instructions.  (We do the same thing on other CPU architectures.)  If someone wants to build libjpeg-turbo with `-march=rv64gc_zvbb`, then that flag will affect all of the non-SIMD code, including the Huffman decoder.  It will be overridden by `-march=rv64gcv` when building the SIMD modules, but since the SIMD modules don't use Zvbb, I don't see a problem.

**Comment by dcommander** [comment]
Would appreciate any feedback that the community can give regarding the proposed optimizations in:
https://github.com/libjpeg-turbo/libjpeg-turbo/tree/rvv-dct-opt

I observe a significant compression speedup with GCC 14 and a small compression speedup with Clang 20.  The submitter observes a significant compression speedup with Clang 21.  Neither of us observes a compelling decompression speedup.

Mainly, I'm looking for additional testing and benchmarking on additional hardware so I can gain some confidence that the code is robust and will not regress in performance.

**Comment by dcommander** [comment]
> Would appreciate any feedback that the community can give regarding the proposed optimizations in: https://github.com/libjpeg-turbo/libjpeg-turbo/tree/rvv-dct-opt
> 
> I observe a significant compression speedup with GCC 14 and a small compression speedup with Clang 20. The submitter observes a significant compression speedup with Clang 21. Neither of us observes a compelling decompression speedup.
> 
> Mainly, I'm looking for additional testing and benchmarking on additional hardware so I can gain some confidence that the code is robust and will not regress in performance.

The optimizations have been merged into main, which will eventually become 3.2.1.  Would still greatly appreciate any feedback, as the feature will land in a bug-fix release.
